Mango Markets Integration (#299)
solana-developers/solana-cookbook: docs/integrations/mango.md

  - - meta
    - name: og:description
      content: Mango Markets offers the industry standard for decentralized, cross-margin trading. Learn how to use and build on top of Mango Markets.
  - - meta
    - name: og:image
      content: https://solanacookbook.com/cookbook-sharing-card.png
  - - meta
    - name: og:image:alt
      content: Solana splash card
  - - meta
    - name: twitter:card
      content: summary
  - - meta
    - name: twitter:site
      content: "@solanacookbook"
  - - meta
    - name: twitter:image
      content: "https://solanacookbook.com/cookbook-sharing-card.png"
  - - meta
    - name: robots
      content: index,follow,noodp
  - - meta
    - name: googlebot
      content: index,follow
---

# Mango Markets

Mango provides a single venue to lend, borrow, swap, and leverage 
trade cryptoassets through a powerful risk engine. With lightning-speed
trade execution, up to 5x leverage, near-zero fees, and attractive
interest rates for lenders, the Mango Markets experience competes
head on with centralized exchanges without any compromise.
You can connect to the Mango Program using the Client API libraries.

## How to get a Mango Group

> A mango group is a basket of cross-margined tokens. Each version 

Author: No, the v3 mainnet group has all the tokens, but the v2 mainnet group lacks some of the tokens found in the devnet groups. You can see so [here](https://github.com/blockworks-foundation/mango-client-ts/blob/main/src/ids.json)

of Mango Markets uses a different Mango Group containing different 
tokens. The current v3 group is `mainnet.1`.

<SolanaCodeGroup>
  <SolanaCodeGroupItem title="TS" active>

  <template v-slot:default>

@[code](@/code/mango/load-group/load-group.en.ts)

  </template>

  <template v-slot:preview>

@[code](@/code/mango/load-group/load-group.preview.en.ts)

  </template>
  
  </SolanaCodeGroupItem>
  
</SolanaCodeGroup>

## How to create a Mango Account

A Mango Account is associated with a Mango Group, and it holds your tokens and allows 
you to trade that Group’s markets. You can find the reference [here](https://blockworks-foundation.github.io/mango-client-v3/classes/MangoClient.html#createMangoAccount). 
The Typescript snippet uses the @solana/wallet-adapter packages. In order to create a valid instruction, the Rust snippet uses the mango-v3 repository as a crate.

<SolanaCodeGroup>
  <SolanaCodeGroupItem title="TS" active>
  
  <template v-slot:default>

@[code](@/code/mango/create-account/create-account.en.ts)

  </template>

  <template v-slot:preview>

@[code](@/code/mango/create-account/create-account.preview.en.ts)

  </template>

  </SolanaCodeGroupItem>

>   <SolanaCodeGroupItem title="Rust">

Reviewer: ```suggestion
  <SolanaCodeGroupItem title="Anchor">
```

  <template v-slot:default>

@[code](@/code/mango/create-account/create-account.en.rs)

  </template>

  <template v-slot:preview>

@[code](@/code/mango/create-account/create-account.preview.en.rs)

  </template>

  </SolanaCodeGroupItem>
</SolanaCodeGroup>

## How to deposit USDC into a Mango Account
After creating a mango account, you'll need to fund it with tokens for trading. 
You can find the reference for the deposit method [here](https://blockworks-foundation.github.io/mango-client-v3/classes/MangoClient.html#deposit). 
This snippet uses the @solana/wallet-adapter packages

<SolanaCodeGroup>
  <SolanaCodeGroupItem title="TS" active>

  <template v-slot:default>

@[code](@/code/mango/deposit/deposit.en.ts)

  </template>

  <template v-slot:preview>

@[code](@/code/mango/deposit/deposit.preview.en.ts)  

  </template>
  
  </SolanaCodeGroupItem>
</SolanaCodeGroup>

## How to place a spot order

> Mango uses Serum DEX for spot markets. You can place a spot order by doing this. 

Reviewer: ```suggestion
Mango interacts with Serum Protocol to place spot orders on markets.
```

You can find the reference for the placeSpotOrder function [here](https://blockworks-foundation.github.io/mango-client-v3/classes/MangoClient.html#placeSpotOrder)

<SolanaCodeGroup>
  <SolanaCodeGroupItem title="TS" active>
    
  <template v-slot:default>

@[code](@/code/mango/place-spot-order/place-spot-order.en.ts) 

  </template>

  <template v-slot:preview>

@[code](@/code/mango/place-spot-order/place-spot-order.preview.en.ts)

  </template>
 
  </SolanaCodeGroupItem>
</SolanaCodeGroup>

## How to load bids

> You can load bids from a Market. Mango uses Serum DEX for markets. You can find out more about Serum's markets [here](https://github.com/project-serum/serum-ts/tree/master/packages/serum)

Reviewer: ```suggestion
Mango uses the market information from Serum Protocol to load bids. You can load them directly from Serum to work with on Mango.
```
Let's avoid linking to Serum here. We'll link to the Serum integration once all of that is complete.

<SolanaCodeGroup>
  <SolanaCodeGroupItem title="TS" active>

  <template v-slot:default>

@[code](@/code/mango/load-bids/load-bids.en.ts)

  </template>

  <template v-slot:preview>

@[code](@/code/mango/load-bids/load-bids.preview.en.ts)

  </template>

  </SolanaCodeGroupItem>
</SolanaCodeGroup>

## How to load asks

> You can load asks from a Market. Mango uses Serum DEX for markets. ou can find out more 
> about Serum's markets [here](https://github.com/project-serum/serum-ts/tree/master/packages/serum)

Reviewer: ```suggestion
Mango uses the market information from Serum Protocol to load asks. You can load them directly from Serum to work with on Mango.
```

  - - meta
    - name: description
      content: Mango Markets offers the industry standard for decentralized, cross-margin trading. Learn how to use and build on top of Mango Markets.
  - - meta
    - name: og:description
      content: Mango Markets offers the industry standard for decentralized, cross-margin trading. Learn how to use and build on top of Mango Markets.
  - - meta
    - name: og:image
      content: https://solanacookbook.com/cookbook-sharing-card.png
  - - meta
    - name: og:image:alt
      content: Solana splash card
  - - meta
    - name: twitter:card
      content: summary
  - - meta
    - name: twitter:site
      content: "@solanacookbook"
  - - meta
    - name: twitter:image
      content: "https://solanacookbook.com/cookbook-sharing-card.png"
  - - meta
    - name: robots
      content: index,follow,noodp
  - - meta
    - name: googlebot
      content: index,follow
---

# Mango Markets

Mango provides a single venue to lend, borrow, swap, and leverage 
trade cryptoassets through an on-chain risk engine.
You can connect to Mango's on-chain program using the Client API libraries.
You'll also need the Solana javascript API library.

<SolanaCodeGroup>

>   <SolanaCodeGroupItem title="TS" active>

Reviewer: Since there is not really a preview here, pressing the switcher button breaks your example. 

Try using `CodeGroup` instead like it is done in [token](https://github.com/solana-developers/solana-cookbook/blob/master/docs/references/token.md)
